### PR TITLE
New all content finder UI: Separate date fields

### PR DESCRIPTION
**Pull request by csutter**
This updates the new all content finder UI to use the canonical GOV.UK Design System "three fields for dates" approach for the date filter/facet, while not breaking existing finders that will continue to use the old UI, and maintaining backwards compatibility in terms of query parameters.

Broadly, this PR:
* introduces a new `DateHashParser` to complement the existing `DateStringParser` (renamed from `DateParser`), and a new `DateParser` that delegates to either of the two based on the type of query parameter
* amends the date input handling for both old and new UI to be able to handle the respective other type (by adding the ability to `DateInput` to convert to strings/hashes regardless of the original input)
* changes the new UI's date facet to use Publishing Components' `date_input` component, via a helper to DRY things up
* tries to clean up date code a little (for example by extracting `DateInput` into a file of its own) and adds some extra testing

### Before
image

### After
image

### CODE REVIEW

File: spec/parsers/date_parser_spec.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+require "spec_helper"
+
+describe DateParser do
+  subject(:date_parser) { described_class.new(date_param) }
+
+  let(:date_param) { "I act like it's my birthday, every day" }
+  let(:date_string_parser) { instance_double(DateStringParser, parse: Date.new(1989, 12, 13)) }
+
+  before do
+    allow(DateStringParser).to receive(:new).and_return(date_string_parser)
```
Comment (kevindew) [inline spec/parsers/date_parser_spec.rb:10]: I would have thought we don't need to mock this and use doubles and instead can call original and assert it was called?

**Review by kevindew — COMMENTED** [review]
Looks great - nice to see consistency with the design system and PR was really well broken up into commits so very pleasant to review. 

My suggestions are mostly on the minor side, not sure I caught any actual problems.

File: spec/parsers/date_parser_spec.rb (hunk 1 above)
Comment (csutter) [reply spec/parsers/date_parser_spec.rb:10]: Fair – the simplest thing is probably just to accept the coupling altogether and test the return value, then we don't even need to assert. Given how strong a dependency there is between them all I doubt it's too much of an issue!

File: app/parsers/date_hash_parser.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  def year
+    value = date_hash.fetch(:year, Time.zone.today.year)
+
+    case value
+    when 0..MILLENNIUM_CUTOFF
+      2000 + value
+    when MILLENNIUM_CUTOFF..99
+      1900 + value
+    else
+      value
```
[the remark this answers is not in the document]
Comment (csutter) [reply app/parsers/date_hash_parser.rb:62]: The current string parsing logic is very lenient towards years, allowing pretty arbitrary years - I'd lean towards just keeping that consistent for now and then reevaluating if we want to be stricter for both.

File: app/models/date_input.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  def date
+    @date ||= DateParser.new(original_input).parse
+  end
+
+  def to_param
+    original_input
+  end
+
+  def to_iso8601
+    date.iso8601
```
[the remark this answers is not in the document]
Comment (csutter) [reply app/models/date_input.rb:20]: The previous code didn't, but it'll be nice for consistency with the new methods – will do 👍

File: app/parsers/date_parser.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class DateParser
+  def initialize(date_param)
+    @date_param = date_param
+  end
+
+  def parse
+    DateStringParser.new.parse(date_param)
```
[the remark this answers is not in the document]
Comment (csutter) [reply app/parsers/date_parser.rb]: Indeed – I was tempted to refactor the existing `Date(String)Parser` anyway as that is very messy with `#parse` taking a parameter that should really be an instance variable (and lots of methods taking the exact same parameter). Might as well do that while I'm here!

File: app/models/date_input.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# resulting from the standard three-field date input.
+class DateInput
+  attr_reader :original_input
+
+  def initialize(original_input)
+    @original_input = original_input
+  end
+
+  def date
+    @date ||= DateParser.new(original_input).parse
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/models/date_input.rb:9]: Nit: `||` is redundant here

File: spec/validators/param_validator_spec.rb [hunk 6]
```diff
@@ -1,31 +1,23 @@
 require "spec_helper"
 
 describe ParamValidator do
```
Comment (kevindew) [inline spec/validators/param_validator_spec.rb:3]: Thanks 👍

**Review by kevindew — APPROVED** [review]
Nice one, looks good to me. Just a tiny thing to fix.

**Pull request merged**